### PR TITLE
[stdlib] Decidable instance for negation

### PR DESCRIPTION
**Pull request by liyishuai**
**Kind:** feature.

- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: theories/Classes/DecidableClass.v
Comment (SkySkimmer) [inline theories/Classes/DecidableClass.v:71]: There was an attempt to stop using `Proof with`/`...`, it may be good to avoid introducing new ones.

**Comment by anton-trunov** [comment]
Looks like this PR needs a review from @coq/typeclasses-maintainers.

**Comment by liyishuai** [comment]
This PR seems to be yet another instance of #7368

**Comment by anton-trunov** [comment]
Are there any takers for this PR?

**Comment by Zimmi48** [comment]
It looks like the @coq/typeclasses-maintainers team would benefit from some additional members. Maybe @CohenCyril?
In the case of this PR, since it is about library file, any typeclass expert could weigh in (not just @CohenCyril but also @Blaisorblade, @robbertkrebbers, etc.)

File: theories/Classes/DecidableClass.v
Comment (Blaisorblade) [inline theories/Classes/DecidableClass.v:73]: #7368 has open questions about `Decidable` — but setting that aside, the bigger question is making sure that instances don't overly complicate inference. This one is probably fine. `Classes/Init.v` even makes `not` TC opaque — is that imported here somehow?

Some (premature?) low-level comments:
1. Instead of `H0`, it's best to use explicitly-introduced names.
2. Other obligations here suggest this proof might profitably be lifted to a lemma (not sure).

Comment (Blaisorblade) [reply theories/Classes/DecidableClass.v:73]: Testing (below) shows `Classes/Init` is required here, and that should be good enough for TC opaqueness!

```coq
About Unconvertible.
About Classes.Init.Unconvertible.
(*
Unconvertible not a defined object.
Init.Unconvertible : forall A : Type, A -> A -> Set

Init.Unconvertible is not universe polymorphic
Arguments Init.Unconvertible _%type_scope
Init.Unconvertible is transparent
Expands to: Constant Coq.Classes.Init.Unconvertible
*)
```

**Comment by Blaisorblade** [comment]
So I think this LGTM (modulo the `H0` nit)? I'll think about it.

**Comment by Zimmi48** [comment]
After talking about this PR in the Coq Call yesterday, I don't think we will get feedback from @coq/typeclasses-maintainers, so I suggest relying on the expertise of @Blaisorblade and having an assignee from @coq/stdlib-maintainers.

Should we create a new maintainer team for typeclasses library files?

**Comment by anton-trunov** [comment]
> Should we create a new maintainer team for typeclasses library files?

I think it's a good idea!

**Comment by Zimmi48** [comment]
@Blaisorblade Would you be interested to be part of such a maintainer team (of typeclasses library files) if we created it? What about other typeclasses users / experts like @CohenCyril @spitters @robbertkrebbers @RalfJung?

**Comment by Blaisorblade** [comment]
I’d accept as long as some other experts do as well, I think I’m still somewhat junior on the topic. I know @Janno has more experience 😉 .

I assume CI failures are spurious, since this did pass at first — so let’s find and finish reviewing I guess.

**Comment by robbertkrebbers** [comment]
> @Blaisorblade Would you be interested to be part of such a maintainer team (of typeclasses library files) if we created it? What about other typeclasses users / experts like @CohenCyril @spitters @robbertkrebbers @RalfJung?

I could, although I'm very limited on time to be honest.

Also, I'm not an expert of the type classes used in the Coq stdlib. In std++ and Iris, we have mostly our own hierarchy of type classes, and only use the classes related to the setoid mechanism like `Proper`, `Reflexive`, `Transitive`, `PreOrder`, etc. from the Coq stdlib.

**Review by Blaisorblade — APPROVED** [review]

**Comment by Blaisorblade** [comment]
> So I think this LGTM (modulo the `H0` nit)? I'll think about it.

I think it's an LGTM.
Other reviewers are not rushing here, so I'd try to make progress. @liyishuai, can you look into the failures and confirm they're spurious? Not sure how/who can restart jobs.

**Comment by Zimmi48** [comment]
> can you look into the failures and confirm they're spurious?

I confirm that these are spurious failures that we've seen everywhere recently (although the macOS one is now fixed if you rebase on `master` but I don't think it's needed for this PR which has no reason to break on macOS in particular).

>Not sure how/who can restart jobs.

In this case, restarting wouldn't help, but in the case of non-deterministic failures, anyone in `@coq/contributors` (e.g., you) can restart a job from the GitHub Checks tab by clicking the re-run buttons (support for this in the case of GitLab CI builds reported by coqbot dates from yesterday actually, and is only applicable to jobs that ran after this date):

2020-11-11-205844_781x476_scrot

**Comment by anton-trunov** [comment]
@Blaisorblade Thanks for looking into this. I'll merge this PR in a couple of days unless someone objects.

**Comment by anton-trunov** [comment]
@coqbot: merge now

**Pull request merged**
